File: java/TrappingRainWater.java
/*
 * Filename: TrappingRainWater.java
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Thu 09 Apr 2015 11:06:14 AM EDT
 */

public class Solution {
	public int trap(int[] A) {
		if (A == null || A.length < 3) {
			return 0;
		}

		int trap = 0;
		int peak = -1;
		int area = 0;

		for (int i = 0; i < A.length; i++) {
			if (isPeak(A, i)) {
				if (peak == -1) {
					peak = i;
				} else if (A[i] >= A[peak]) {
					trap += area;
					area = 0;
					peak = i;
				} else {
					area += Math.max(0, A[peak] - A[i]);
				}
			} else {
				if (peak != -1) {
					area += Math.max(0, A[peak] - A[i]);
				}
			}
		}

		peak = A.length;
		area = 0;

		for (int i = A.length - 1; i >= 0; i--) {
			if (isPeak(A, i)) {
				if (peak == A.length) {
					peak = i;
				} else if (A[i] > A[peak]) {
					trap += area;
					area = 0;
					peak = i;
				} else {
					area += Math.max(0, A[peak] - A[i]);
				}
			} else {
				if (peak != A.length) {
					area += Math.max(0, A[peak] - A[i]);
				}
			}
		}

		return trap;
	}

	private boolean isPeak(int[] A, int i) {
		return (i == 0 && A[i] >= A[i+1]) ||
			(i == A.length - 1 && A[i] >= A[i-1]) ||
			(i != 0 && i != A.length -1 && A[i] >= A[i-1] && A[i] >= A[i+1]);
	}
}

File: cpp/trapping_rain_water.cc
/*
 * Filename: trapping_rain_water.cc
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Sun 26 Jul 2015 11:07:31 AM EDT
 */

#include <iostream>
#include <vector>

using namespace std;

class Solution {
	public:
		int trap(vector<int>& height) {
			if (!height.size()) {
				return 0;
			}

			vector<int>::size_type i = 0;
			vector<int>::size_type j = height.size() - 1;
			int secH = 0;
			int res = 0;

			while (i < j) {
				if (height[i] < height[j]) {
					secH = max(secH, height[i]);
					res += secH - height[i];
					i++;
				} else {
					secH = max(secH, height[j]);
					res += secH - height[j];
					j--;
				}
			}

			return res;
		}
};
